codec/src/main/java/io/netty/handler/codec/DelimitedLengthFieldBasedFrameDecoder.java

```diff
@@ -0,0 +1,217 @@
+ * under the License.
+ */
+package io.netty.handler.codec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.serialization.ObjectDecoder;
+
+import java.nio.charset.Charset;
+import java.util.List;
+
+/**
+ * Frames input by looking for a length field that is delimited, as opposed to being a fixed length
+ *
+ * Uses the same approach as the {@link DelimiterBasedFrameDecoder} to find delimiter in input
+ */
+public class DelimitedLengthFieldBasedFrameDecoder extends ByteToMessageDecoder {
+
+  private final int maxFrameLength;
+  private final int lengthAdjustment;
+  private final boolean failFast;
+  private final Charset lengthFieldCharset;
+  private final ByteBuf delimiter;
+  private final boolean trimLengthString;
+
+  private boolean discardingTooLongFrame;
+  private long tooLongFrameLength;
+  private long bytesToDiscard;
+  private boolean consumingLength = true;
+  private long frameLength;
+
+  public DelimitedLengthFieldBasedFrameDecoder(
+      int maxFrameLength,
+      int lengthAdjustment,
+      boolean failFast,
+      ByteBuf delimiter,
+      Charset lengthFieldCharset,
+      boolean trimLengthString
+  ) {
+    if (delimiter == null) {
+      throw new NullPointerException("delimiter");
+    }
+    if (!delimiter.isReadable()) {
+      throw new IllegalArgumentException("empty delimiter");
+    }
+
+    this.delimiter = delimiter.slice(delimiter.readerIndex(), delimiter.readableBytes());
+    this.lengthFieldCharset = lengthFieldCharset;
+
+    this.maxFrameLength = maxFrameLength;
+    this.lengthAdjustment = lengthAdjustment;
+    this.failFast = failFast;
+    this.trimLengthString = trimLengthString;
+  }
+
+  @Override
+  protected final void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+    Object decoded = decode(ctx, in);
+    if (decoded != null) {
+      out.add(decoded);
+    }
+  }
+
+  /**
+   * Create a frame out of the {@link ByteBuf} and return it.
+   *
+   * @param ctx the {@link ChannelHandlerContext} which this {@link ByteToMessageDecoder} belongs to
+   * @param in the {@link ByteBuf} from which to read data
+   * @return frame           the {@link ByteBuf} which represent the frame or {@code null} if no frame could
+   * be created.
+   */
+  protected Object decode(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
+    if (discardingTooLongFrame) {
+      long bytesToDiscard = this.bytesToDiscard;
+      int localBytesToDiscard = (int) Math.min(bytesToDiscard, in.readableBytes());
+      in.skipBytes(localBytesToDiscard);
+      bytesToDiscard -= localBytesToDiscard;
+      this.bytesToDiscard = bytesToDiscard;
+
+      failIfNecessary(false);
+    }
+
+    if (consumingLength) {
+      int delimIndex = indexOf(in, delimiter);
+      if (delimIndex < 0) {
+        return null;
+      }
+
+      final String lengthStr = in.toString(in.readerIndex(), delimIndex, lengthFieldCharset);
```

Comment:
ditto. Consider extracting a getting of the `frameLength` into a separate protected method.